### PR TITLE
make Dimension be compatible with integer

### LINKED ISSUE #11974: tf.reshape does not accept Dimension objects for the shape parameter
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 x64
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**:  1.2.1
- **Python version**: 3.6.1 (Anaconda 4.4.0 64-bit)
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: 
- **GPU model and memory**: GTX 780
- **Exact command to reproduce**: tf.reshape()

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

tf.reshape does not accept a list with mixed Integer and Dimension() as elements for the shape parameter. It should accept shapes that have Dimension as elements since tensor shapes consist of dimensions. Specifically, tf.tensor.shape returns a list of Dimensions, therefore using a similar object to specify a shape in tf.reshape should not cause an error.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

Example code:

```
X = tf.placeholder("float", [1,784])
X = tf.reshape(X, [1, X.shape[-1]])
```

results in

```
TypeError: Failed to convert object of type <class 'list'> to Tensor. Contents: [1, Dimension(784)]. Consider casting elements to a supported type.
```

This is rectified by casting X.shape[-1] to int before passing to tf.reshape.

A similar unrelated issue is that X.shape returns (784,) while it should return (1,784). This requires reshaping to turn the placeholder back into a 2D tensor. I haven't determined if this is a bug but it occurs when the tensor is explicitly specified as a 2D tensor so it is probably worth changing.

### PR DESCRIPTION
**Pull request by facaiy**
The PR is another solution to fix #11974 .  I believe that it is better than #12127 .

### What changes were proposed in this pull request?

Make Dimension to be compatible with integer, so `[1, Dimension(2)]` will be casted to `[1, 2]` automatically.

### How was this patch tested?

+ [x] add a doctest. 
+ [x] add an unit test.
+ [x] pass all unit tests.

### CODE REVIEW

**Comment by facaiy** [comment]
cc @yaroslavvb @pmccarter @martinwicke , who I believe are interested in this PR. Could you take a look please?

File: tensorflow/python/framework/tensor_shape.py [hunk 1]
```diff
@@ -428,6 +429,8 @@ def __init__(self, dims):
       TypeError: If dims cannot be converted to a list of dimensions.
     """
     # TODO(irving): Eliminate the single integer special case.
+    from tensorflow.python.util import compat
```
Comment (alextp) [inline tensorflow/python/framework/tensor_shape.py]: imports should be at the top of the file in tensorflow

File: tensorflow/python/util/compat.py [hunk 2]
```diff
@@ -40,6 +40,7 @@
 import numpy as _np
 import six as _six
 
+from tensorflow.python.framework import tensor_shape
```
Comment (alextp) [inline tensorflow/python/util/compat.py]: I don't like this dependency. util should be small utilities which do not depend on any of the concrete tensorflow stuff.

I think the other fix is better.

File: tensorflow/python/framework/tensor_shape.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/framework/tensor_shape.py]: `tensor_shape` and `compat` are depended by each other if imports are at the top. I don't know how to deal with the circular imports.

File: tensorflow/python/util/compat.py (hunk 2 above)
Comment (facaiy) [reply tensorflow/python/util/compat.py]: Thanks for comment. Could you please give me a bit more detail about where to fix?

How about hack [`def _FilterInt(v):`](https://github.com/tensorflow/tensorflow/blob/c48aaadbe9c5802316d73e40bbcde34221277b81/tensorflow/python/framework/tensor_util.py#L238) function? 
or add a `if` branch at [`_AssertCompatible(values, dtype)`](https://github.com/tensorflow/tensorflow/blob/c48aaadbe9c5802316d73e40bbcde34221277b81/tensorflow/python/framework/tensor_util.py#L383)?

Comment (alextp) [reply tensorflow/python/util/compat.py]: Essentially, I do not think that this approach is the right way to fix the problem. compat.as_int shouldn't know anything about dimension. I think something like the change made in  #12558  is the right way to go.

I'll close this pull request now.

Comment (facaiy) [reply tensorflow/python/util/compat.py]: Hi, @alextp . Perhaps the [new commit](https://github.com/facaiy/tensorflow/commits/BUG/tf_reshape_mix_int_with_dim) will be interesting for you. Unfortunately, the PR has been closed.

Comment (facaiy) [reply tensorflow/python/util/compat.py]: How about #12127 ? It is more straightforward.

File: tensorflow/core/ops/ops.pbtxt [hunk 3]
```diff
@@ -19893,7 +19893,7 @@ op {
...
...\'t\' is [[[1, 1], [2, 2]],\n#                [[3, 3], [4, 4]]]\n# tensor \'t\' has shape [2, 2, 2]\nreshape(t, [2, 4]) ==> [[1, 1, 2, 2],\n                        [3, 3, 4, 4]]\n\n# tensor \'t\' is [[[1, 1, 1],\n#                 [2, 2, 2]],\n#                [[3, 3, 3],\n#                 [4, 4, 4]],\n#                [[5, 5, 5],\n#                 [6, 6, 6]]]\n# tensor \'t\' has shape [3, 2, 3]\n# pass \'[-1]\' to flatten \'t\'\nreshape(t, [-1]) ==> [1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 6]\n\n# -1 can also be used to infer the shape\n\n# -1 is inferred to be 9:\nreshape(t, [2, -1]) ==> [[1, 1, 1, 2, 2, 2, 3, 3, 3],\n                         [4, 4, 4, 5, 5, 5, 6, 6, 6]]\n# -1 is inferred to be 2:\nreshape(t, [-1, 9]) ==> [[1, 1, 1, 2, 2, 2, 3, 3, 3],\n                         [4, 4, 4, 5, 5, 5, 6, 6, 6]]\n# -1 is inferred to be 3:\nreshape(t, [ 2, -1, 3]) ==> [[[1, 1, 1],\n                              [2, 2, 2],\n                              [3, 3, 3]],\n                             [[4, 4, 4],\n                              [5, 5, 5],\n                              [6, 6, 6]]]\n\n# tensor \'t\' is [7]\n# shape `[]` reshapes to a scalar\nreshape(t, []) ==> 7\n```"
```
Comment (alextp) [inline tensorflow/core/ops/ops.pbtxt]: Revert this change? This is a generated file.

**Comment by alextp** [comment]
I am ready to approve this when the change to ops.pbtxt gets reverted.

**Comment by alextp** [comment]
Jenkins, test this please.

**Review by alextp — APPROVED** [review]

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by facaiy** [comment]
Thanks. The new commits are modified on my personal laptop. It is no surprise if some unit tests are broken. If so, I'll fix them on next Monday.

**Comment by facaiy** [comment]
It seems a trivial mistake, and has been fixed. Could you retest the PR? Thanks.

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by facaiy** [comment]
Thanks, @alextp .  All tests pass.

**Pull request merged**